src/System.Threading.Channels/src/System/Threading/Channels/BoundedChannel.cs

```diff
@@ -168,9 +170,17 @@ public override bool TryComplete(Exception error)
 
                     // Mark that we're done writing.
                     parent._doneWriting = error ?? ChannelUtilities.s_doneWritingSentinel;
+
+                    ChannelUtilities.FailInteractors<WriterInteractor<T>, VoidResult>(parent._blockedWriters, ChannelUtilities.CreateInvalidCompletionException(error));
+                    ChannelUtilities.WakeUpWaiters(ref parent._waitingWriters, result: false, error: error);
+                    parent._waitingReaders = null;
+
                     completeTask = parent._items.IsEmpty;
                 }
 
+                // Avoid waking the readers inside the lock as the readers will be invoked synchronously.
```

Comment:
> "will be invoked synchronously"

They may or may not be; it depends on whether AllowSynchronousContinuations was specified.  If it was, then they will be; if not, they won't be.